### PR TITLE
Add option to open the datepicker at focus in swirl-date-input

### PR DESCRIPTION
**Pull request by joseabernardes**
## Summary

- [Ticket](#)
- [Design](#)
- [Storybook](#)

### General

- [x] [Changeset added](https://github.com/changesets/changesets/blob/main/docs/intro-to-using-changesets.md)
- [x] The submitted code is organized and formatted according to our [💅 Style Guide](https://swirl-storybook.flip-app.dev/?path=/docs/requirements-style-guide--docs).
  - [ ] The code is formatted with Prettier.
  - [ ] There are no linting errors.

### For new or updated components

https://swirl-storybook.flip-app.dev/?path=/docs/contributions-merge-publish--page

- [ ] The changes do not contain any breaking changes.
- [ ] The changes do not introduce new components that don't belong in the library (e.g. non-reusable components, highly specialized components, components including business logic)
- [ ] The component documentation is updated.
- [ ] The changes meet the [🤖 testing requirements](https://swirl-storybook.flip-app.dev/?path=/docs/requirements-testing--docs).
  - [ ] New features are tested.
  - [ ] In case of bug fixes, regression tests have been added.
  - [ ] All tests are 🟢.
- [ ] The changes meet the [🧏‍♀️ accessibility requirements](https://swirl-storybook.flip-app.dev/?path=/docs/requirements-accessibility--docs).
  - [ ] WCAG 2.1 Level A and Level AA requirements are met.
  - [ ] The Storybook a11y addon shows no errors.
  - [ ] The changes have been tested with a screen reader.
  - [ ] Keyboard controls have been tested, if applicable.
  - [ ] Components implementing form controls (inputs, buttons, selects, etc.) do not use Shadow DOM, and instead use Stencil's scoping mechanism
- [ ] The changes use our [🌈 theming concept](https://swirl-storybook.flip-app.dev/?path=/docs/requirements-theming--docs).
  - [ ] Design tokens have been used where appropriate.
  - [ ] The component has been visually checked in combination with the "Light" and "Dark" theme.
- [ ] The changes meet our [🌍 internationalization requirements](https://swirl-storybook.flip-app.dev/?path=/docs/requirements-internationalization--docs).
  - [ ] No static text is used.
  - [ ] The component doesn't break with longer texts or different text wrappings.
  - [ ] Number, currency and date values can be formatted appropriately.
- [ ] The changes work in all supported browsers and viewports. See [📱 Responsive Design](https://swirl-storybook.flip-app.dev/?path=/docs/requirements-responsive-design--docs)

### CODE REVIEW

File: packages/swirl-components/src/components/swirl-date-input/swirl-date-input.tsx [hunk 1]
```diff
@@ -46,6 +46,7 @@ export class SwirlDateInput {
   @Prop() labels?: WCDatepickerLabels;
   @Prop() locale?: string = "en-US";
   @Prop() placeholder?: string = "yyyy-mm-dd";
+  @Prop() preferredInputMode?: "input" | "pick" = "pick";
```
Comment (Sqrrl) [inline packages/swirl-components/src/components/swirl-date-input/swirl-date-input.tsx:49]: Issue: To not make this a breaking change, the default should be "input".

**Review by Sqrrl — CHANGES_REQUESTED** [review]
Issue: When I tab into the form control, the picker opens correctly. When I tab again to interact with the picker, the picker closes.

File: packages/swirl-components/src/components/swirl-date-input/swirl-date-input.tsx (hunk 1 above)
Comment (joseabernardes) [reply packages/swirl-components/src/components/swirl-date-input/swirl-date-input.tsx:49]: Sure, it was just to help me test on mobile :P

**Comment by joseabernardes** [comment]
> Issue: When I tab into the form control, the picker opens correctly. When I tab again to interact with the picker, the picker closes.

Excellent point!
I was investigating, and it appears to be an already existing behavior, and the reason for it is because when we give focus to the popover(datepicker) via tab, the popover-trigger(calendar button) gets hidden (because the input itself loses focus) and then the trigger closes the datepicker when we try to navigate in it (via tab or mouse).
One of the solutions that I found was to set the `hidePopoverWhenInvisible` prop to false on the trigger, but that could have unwanted side effects. 
What do you think about this approach?

**Comment by Sqrrl** [comment]
> > Issue: When I tab into the form control, the picker opens correctly. When I tab again to interact with the picker, the picker closes.
> 
> Excellent point! I was investigating, and it appears to be an already existing behavior, and the reason for it is because when we give focus to the popover(datepicker) via tab, the popover-trigger(calendar button) gets hidden (because the input itself loses focus) and then the trigger closes the datepicker when we try to navigate in it (via tab or mouse). One of the solutions that I found was to set the `hidePopoverWhenInvisible` prop to false on the trigger, but that could have unwanted side effects. What do you think about this approach?

We could try and keep the form control "open" when the date picker has focus, so the calendar icon doesn't disappear.

File: packages/swirl-components/src/components/swirl-form-control/swirl-form-control.tsx [hunk 2]
```diff
@@ -173,7 +173,7 @@ export class SwirlFormControl {
   private onKeyDown = (event: KeyboardEvent) => {
     if (event.key === "Tab") {
       setTimeout(() => {
-        if (!this.el.contains(getActiveElement())) {
+        if (!isDescendantOf(getActiveElement(), this.el)) {
```
Comment (joseabernardes) [inline packages/swirl-components/src/components/swirl-form-control/swirl-form-control.tsx:176]: The issue with the tab navigation was due to this.
As the popover was inside a Shadow Root, it wasn't considered a descendent of the control itself, so the input would lose focus here.
I then implemented the `isDescendantOf ` utility method which checks whether an element is a descendant of another, even if it is inside a Shadow Root.

File: packages/swirl-components/src/components/swirl-date-input/swirl-date-input.spec.tsx [hunk 3]
```diff
@@ -1,16 +1,34 @@
 const maskSpy = jest.fn();
```
Comment (Sqrrl) [inline packages/swirl-components/src/components/swirl-date-input/swirl-date-input.spec.tsx:1]: Issue: The tests are technically green, but running them locally `yarn stencil:test:watch` produces some errors caused by the setTimeout usage in focus().

**Pull request merged**